server/mlabns/util/maxmind.py

```diff
@@ -20,34 +26,76 @@ def __init__(self, city=None, country=None, latitude=None, longitude=None):
         self.longitude = longitude
 
 
-def get_ip_geolocation(remote_addr):
+def get_ip_geolocation(remote_addr,
+        db_type=constants.GEOLOCATION_MAXMIND_USE_FILE):
     """Returns the geolocation data associated with an IP address.
 
     Args:
         remote_addr: A string describing an IPv4 or IPv6 address.
+        db_type: An indicator whether to use the Mamind datastore or file.
 
     Returns:
         A GeoRecord if the geolocation data is found in the db,
         otherwise an empty GeoRecord.
     """
-    try:
-        return get_ipv4_geolocation(remote_addr)
-    except ipaddr.AddressValueError:
-        pass
 
-    try:
-        return get_ipv6_geolocation(remote_addr)
-    except ipaddr.AddressValueError:
-        pass
+    if db_type == constants.GEOLOCATION_MAXMIND_USE_FILE:
+        try:
+            return get_ip_geolocation_datafile(remote_addr)
+        except socket.error:
+            pass
+    elif db_type == constants.GEOLOCATION_MAXMIND_USE_DATASTORE:
+        try:
+            return get_ipv4_geolocation_datastore(remote_addr)
+        except ipaddr.AddressValueError:
+            pass
+
+        try:
+            return get_ipv6_geolocation_datastore(remote_addr)
+        except ipaddr.AddressValueError:
+            pass
 
     # Return an empty GeoRecord.
     logging.warning('Returning empty record')
     return GeoRecord()
 
-def get_ipv4_geolocation(remote_addr,
-                         ipv4_table=model.MaxmindCityBlock,
-                         city_table=model.MaxmindCityLocation):
-    """Returns the geolocation data associated with an IPv4 address.
+def get_ip_geolocation_datafile(remote_addr,
+        city_file=constants.GEOLOCATION_MAXMIND_CITY_FILE):
+    """Returns the geolocation data associated with an IPv4 address retrieved
+    from the Maxmind database through the pygeoip module.
+
+    Args:
+        remote_addr: A string describing an IP address (v4 or v6).
+        city_file: The path to the Maxmind .dat file.
+
+    Returns:
+        A GeoRecord containing the geolocation data if is found in the db,
+        otherwise an empty GeoRecord.
+
+    Raises:
+        socket.error, if 'remote_addr' is not a valid IP address.
+    """
+    geo_record = GeoRecord()
+
+    geo_city_block = pygeoip.GeoIP(city_file,
+        flags=pygeoip.const.MEMORY_CACHE).record_by_addr(remote_addr)
```

Comment:
Can pygeoip preserve its cache from request to request on AppEngine? We might be better off with `GEOIP_STANDARD`.